### DISCUSSION TITLE
About RearchConsumer and StatelessWidget

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#161] [General]
So, In my way back to `rearch_flutter` I had a problem for 30 minutes trying to figuring out why a `RearchConsumer` was not refreshing, and it was because I was passing it data on its constructor (like with State(less/ful)Widgets).

So my question is what would be the preferred way to handle a situation when you need a `RearchConsumer` but also you need to pass it some values (that could change over the time) via its constructor?

Would you advice against that "mixin" even if you know the passed values should not change over the time?

**Comment by busslina** [comment]
I had some thought on this, and could be interesting:

What about allowing a `RearchConsumer` to update when its data properties changes?
Is that feasible? Maybe an extra parameter on `build()` would be needed. One that informs of the changes between the old and the new values

**Reply by busslina** [reply]
I predict your answer is that it goes against ReArch philosphy, but I want to know your thoughts on it.

**Reply by busslina** [reply]
It seems to need to split the logic into widget and state, like StatefulWidget

**Comment by GregoryConrad** [comment]
How was the RearchConsumer “not refreshing” when you passed it new constructor parameters? In order to get the new constructor parameters it’d have to be rebuilt, wouldn’t it?

It’s completely fine to pass in data via the constructor, and it shouldn’t require any extra handling

**Reply by busslina** [reply]
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/a719d621-b544-4b95-a41f-5a96115a0d49)

So just by changing from `StatelessWidget` to `RearchConsumer` it stops rebuilding correctly. As you can see, `Header.build()` is missing after `Header.constructor()`.

### header.dart
```dart
abstract class Header extends RearchConsumer {
  Header({
    super.key,
    required this.parentHeight,
    required this.title,
    required this.subtitle,
    required this.onTitleTap,
    this.icon,
  }) {
    print('Header.constructor() -- Title: $title');
  }

  static const double _minHeight = 85;

  final double parentHeight;
  final String title;
  final String? subtitle;
  final void Function() onTitleTap;
  final Widget? icon;

  String? get fontFamily => null;

  Color? get bgColor => null;

  @override
  Widget build(BuildContext context, WidgetHandle use) {
    print('Header.build() -- Title: $title');

    final icon = this.icon;

    return Row(
      children: [
        // Icon left
        if (icon != null) icon,

        // Title / Subtitle
        Expanded(child: _buildTitleSubtitle()),

        // Icon right
        if (icon != null) icon,
      ],
    )
        .withSizeConstraints(
          BoxConstraints(
            minHeight: _minHeight,
            maxHeight: max(_minHeight, parentHeight / 10),
          ),
        )
        .bgColor(Colors.purple);
  }

  Widget _buildTitleSubtitle() => Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          _buildTitle(),
          if (subtitle != null) _buildSubtitle().marginTop(5),
        ],
      );

  Widget _buildTitle() => FittedBox(
        child: Label(
          title,
          color: Colors.lime,
          fontSize: 32,
          fontFamily: fontFamily,
          textAlign: TextAlign.center,
          onTap: onTitleTap,
        ),
      );

  Widget _buildSubtitle() => Label(
        subtitle!,
        color: Colors.lime,
        fontSize: 18,
        textAlign: TextAlign.center,
      );
}
```

### Screen.buildHeader()
```dart
@override
  wlib.Header buildHeader(
    BuildContext context,
    CapsuleReader use, {
    required double parentHeight,
  }) {
    final language = use(languageCapsule);

    print('Screen.buildHeader() -- Title: ${language.title}');

    return Header(
      parentHeight: parentHeight,
      title: language.title,
      subtitle: subtitle,
      onTitleTap: () => context.popToNamed(Routes.index),
      icon: Icons.virus.onTap(() => context.popToNamed(Routes.index)),
    );
  }
```

**Reply by GregoryConrad** [reply]
At first I thought this was something to do with your Header being an `abstract class` (and a build hiding somewhere in the inheritance), but I can confirm I've got this reproducing with:

```dart
class Parent extends StatefulWidget {
  const Parent({super.key});

  @override
  State<Parent> createState() => _ParentState();
}

class _ParentState extends State<Parent> {
  int count = 0;
  @override
  Widget build(BuildContext context) {
    return TextButton(
      onPressed: () => setState(() => count++),
      // child: Child1(count),
      child: ChildConsumer(count),
    );
  }
}

class Child1 extends StatelessWidget {
  const Child1(this.i, {super.key});
  final int i;
  @override
  Widget build(BuildContext context) {
    return Text('$i');
  }
}

class ChildConsumer extends RearchConsumer {
  const ChildConsumer(this.i, {super.key});
  final int i;
  @override
  Widget build(BuildContext context, WidgetHandle use) {
    print('$i');
    return Text('$i');
  }
}
```

So something about the underlying RearchElement. Not sure if I just missed a method I was supposed to override, or something in the Flutter framework is broken. Will take a look.

**Reply by GregoryConrad** [reply]
@busslina fixed in flutter_rearch 1.6.1. Updating your dependencies should fix the issue.

**Reply by busslina** [reply]
https://github.com/GregoryConrad/rearch-dart/issues/163#issuecomment-2082862831

**Reply by busslina** [reply]
Now it's working okay !!
